### PR TITLE
claper: init at 1.1.1

### PR DESCRIPTION
**Pull request by drupol**
Hello,

I'm looking for using [Claper](https://github.com/ClaperCo/Claper) webapp, sadly, it's not available in nixpkgs yet.

I have started to do something, but It's obviously not working yet.

This PR is to gather some help from the community, and in the end provide Claper in Nix.

Tasks:

- [x] Initial `default.nix` file
- [x] Fetch the source and compile it
- [ ] Deal with `npm` stuff
- [ ] Make sure we can run the service in NixOS
- [ ] Provide the service module configuration

The documentation to install it is here: https://github.com/ClaperCo/Claper#installation

Further documentation: https://docs.claper.co/configuration.html

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/web-apps/claper/default.nix [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    yarnLock = ./yarn.lock;
+  };
+in
+beamPackages.mixRelease {
+  inherit pname version src mixFodDeps;
+
+  nativeBuildInputs = [ nodejs ];
+
+  postBuild = ''
+    export NODE_OPTIONS=--openssl-legacy-provider # required for webpack compatibility with OpenSSL 3 (https://github.com/webpack/webpack/issues/14532)
```
Comment (SuperSandro2000) [inline pkgs/servers/web-apps/claper/default.nix:42]:
```suggestion
    # required for webpack compatibility with OpenSSL 3 (https://github.com/webpack/webpack/issues/14532)
    export NODE_OPTIONS=--openssl-legacy-provider
```

File: nixos/modules/services/web-apps/claper.nix [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        claper-postgres = {
+          after = [ "postgresql.service" ];
+          partOf = [ "claper.service" ];
+          serviceConfig = {
+            Type = "oneshot";
+            User = config.services.postgresql.superUser;
+            RemainAfterExit = true;
+          };
+          ensureUsers = [ "claper" ];
+          ensureDatabases = [ cfg.database.postgres.dbname ];
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/claper.nix:126]: I think you want to ensure permissions, too.

File: nixos/modules/services/web-apps/claper.nix [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+            ENDPOINT_HOST = cfg.server.endpoint_host;
+            ENDPOINT_PORT = cfg.server.endpoint_port;
+            PORT = cfg.server.port;
+          };
+
+          path = [ pkgs.claper ]
+            ++ optional cfg.database.postgres.setup config.services.postgresql.package;
+          script = ''
+            exec claper start
+          '';
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/claper.nix:102]: This should be ExecStart

File: nixos/modules/services/web-apps/claper.nix [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+              "postgresql.service"
+              "claper-postgres.service"
+            ];
+
+          environment = {
+            DATABASE_URL = cfg.database.postgres.url;
+            SECRET_KEY_BASE = cfg.secret_key_base;
+            ENDPOINT_HOST = cfg.server.endpoint_host;
+            ENDPOINT_PORT = cfg.server.endpoint_port;
+            PORT = cfg.server.port;
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/claper.nix:95]: We should use socket authentication

File: nixos/modules/services/web-apps/claper.nix [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    };
+
+    server = {
+      secretKeybaseFile = mkOption {
+        type = types.either types.path types.str;
+        description = lib.mdDoc ''
+          Path to the secret used by the `phoenix`-framework. Instructions
+          how to generate one are documented in the
+          [
+          framework docs](https://hexdocs.pm/phoenix/Mix.Tasks.Phx.Gen.Secret.html#content).
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/claper.nix:38]:
```suggestion
          [framework docs](https://hexdocs.pm/phoenix/Mix.Tasks.Phx.Gen.Secret.html#content).
```

File: nixos/modules/services/web-apps/claper.nix [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+          type = types.str;
+          description = lib.mdDoc ''
+            Name of the database to use.
+          '';
+        };
+        url = mkOption {
+          default = "postgres://claper:claper@127.0.0.1:5432/claper";
+          type = types.str;
+          description = lib.mdDoc ''
+            Path to the UNIX domain-socket to communicate with `postgres`.
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/claper.nix:25]: the default is not a unix socket path

**Review by SuperSandro2000 — COMMENTED** [review]
The postgres setup is weird. Why not use the standard way to create databases?

File: nixos/modules/services/web-apps/claper.nix (hunk 4 above)
Comment (drupol) [reply nixos/modules/services/web-apps/claper.nix:94]: I tried to use socket but Claper requires an URL.

**Pull request closed without merging**